Code/Engine/GameEngine/StateMachine/Implementation/StateMachine.cpp

```diff
@@ -0,0 +1,462 @@
+      }
+      else
+      {
+        ezLog::Error("Unknown state machine state type '{}'", sTypeName);
+        return EZ_FAILURE;
+      }
+    }
+  }
+
+  // transitions
+  {
+    ezUInt32 uiNumTransitions = 0;
+    stream >> uiNumTransitions;
+
+    for (ezUInt32 i = 0; i < uiNumTransitions; ++i)
+    {
+      ezUInt32 uiFromStateIndex = 0;
+      ezUInt32 uiToStateIndex = 0;
+
+      stream >> uiFromStateIndex;
+      stream >> uiToStateIndex;
+
+      stream >> sTypeName;
+      if (ezRTTI* pType = ezRTTI::FindTypeByName(sTypeName))
+      {
+        ezUniquePtr<ezStateMachineTransition> pTransition = pType->GetAllocator()->Allocate<ezStateMachineTransition>();
+        EZ_SUCCEED_OR_RETURN(pTransition->Deserialize(stream));
+
+        EZ_SUCCEED_OR_RETURN(AddTransition(uiFromStateIndex, uiToStateIndex, std::move(pTransition)));
+      }
+      else
+      {
+        ezLog::Error("Unknown state machine transition type '{}'", sTypeName);
+        return EZ_FAILURE;
+      }
+    }
+  }
+
+  return EZ_SUCCESS;
+}
+
+//////////////////////////////////////////////////////////////////////////
+
+ezStateMachineInstance::ezStateMachineInstance(ezReflectedClass& owner, const ezSharedPtr<const ezStateMachineDescription>& pDescription /*= nullptr*/)
+  : m_owner(owner)
+  , m_pDescription(pDescription)
+{
+  if (pDescription != nullptr && pDescription->m_uiTotalInstanceDataSize > 0)
+  {
+    m_InstanceData.SetCountUninitialized(pDescription->m_uiTotalInstanceDataSize);
+    m_InstanceData.ZeroFill();
+
+    ezUInt32 uiOffset = 0;
+    for (auto pAttribute : pDescription->m_InstanceDataAttributes)
+    {
+      uiOffset = ezMemoryUtils::AlignSize(uiOffset, pAttribute->m_uiTypeAlignment);
+
+      if (pAttribute->m_ConstructorFunction != nullptr)
+      {
+        pAttribute->m_ConstructorFunction(GetInstanceData(uiOffset));
+      }
+
+      uiOffset += pAttribute->m_uiTypeSize;
+    }
+  }
+}
+
+ezStateMachineInstance::~ezStateMachineInstance()
+{
+  ExitCurrentState();
+
+  m_pCurrentState = nullptr;
+  m_uiCurrentStateIndex = ezInvalidIndex;
+
+  if (m_pDescription != nullptr && m_pDescription->m_uiTotalInstanceDataSize > 0)
+  {
+    ezUInt32 uiOffset = 0;
+    for (auto pAttribute : m_pDescription->m_InstanceDataAttributes)
+    {
+      uiOffset = ezMemoryUtils::AlignSize(uiOffset, pAttribute->m_uiTypeAlignment);
+
+      if (pAttribute->m_DestructorFunction != nullptr)
+      {
+        pAttribute->m_DestructorFunction(GetInstanceData(uiOffset));
+      }
+
+      uiOffset += pAttribute->m_uiTypeSize;
+    }
+
+    m_InstanceData.Clear();
+  }
+}
+
+void ezStateMachineInstance::SetState(ezStateMachineState* pState)
+{
+  if (pState != nullptr && m_pDescription != nullptr)
+  {
+    ezUInt32 uiStateIndex = ezInvalidIndex;
+    if (m_pDescription->m_StateNameToIndexTable.TryGetValue(pState->GetNameHashed(), uiStateIndex))
+    {
+      SetStateInternal(uiStateIndex);
+      return;
+    }
+  }
```

Comment:
If we have a description, but can't find the given state, wouldn't that be an error ?